Add mainnet forking to docs (#917)
nomiclabs/buidler: docs/guides/mainnet-forking.md

# Mainnet forking

You can start an instance of the Hardhat Network that forks the mainnet. This means that it will simulate having the same state as the mainnet, but it will work as a local development node. That way you can interact with deployed protocols and test complex interactions without the risk of doing it in the real mainnet.

> To use this feature, the only thing you need is the URL of a node to connect to. This can be an [Infura](https://infura.io/) or [Alchemy](https://alchemyapi.io/) node, or a local one. If you don't know which to use, we recommend you use [Alchemy](https://alchemyapi.io/) 

Reviewer: ```suggestion
To use this feature, you only the URL of a node to connect to. This can be an [Infura](https://infura.io/) or [Alchemy](https://alchemyapi.io/) node, or a local one. If you don't know which one to use, we recommend using [Alchemy](https://alchemyapi.io/) 
```

## Forking from mainnet

The easiest way to test this feature is to start a node from the command line:

```
npx hardhat node --fork https://eth-mainnet.alchemyapi.io/v2/<key>
```

Here you have to replace `<key>` with your Alchemy key. If you are using Infura, you need a different URL, and a project id instead of a key.

You can also configure Hardhat to always do this, by setting it in the config:

```js
networks: {
  hardhat: {
    forking: {
      url: "https://eth-mainnet.alchemyapi.io/v2/<key>"
    }
  }
}
```

## Pinning a block

> If you start a forked node the way we explained in the previous section, Hardhat Network will fork the mainnet from the latest block available at that moment. This has some problems.

Reviewer: Hardhat Network will by default fork from the latest mainnet block. While this might be practical depending on the context, we recommend forking from a specific block number to set up a test suite that depends on forking.

> First, it's not reproducible: you might run it at some point and it will use the block 11095000, and then run it one hour later and it will use the block 11095240. In that interval of time, the state of the blockchain will change, and this could cause your tests or scripts to behave differently.

Reviewer: ```suggestion
There are two reasons for this:
- The state your tests run against may change between runs. This could cause your tests or scripts to behave differently.
```

> A second problem with using the latest block number is performance. Every time the node fetches some data from the mainnet, Hardhat caches it on disk to avoid doing unnecessary requests. But if you restart your node, then this saved data is no longer useful.

Reviewer: ```suggestion
- Pinning enables caching. Every time data is fetched from mainnet, Hardhat Network caches it on disk to speed up future access. If you don't pin the block, there's going to be new data with each new block and the cache won't be useful. We measured up to 20x speed improvements with block pinning.
```

> The solution to both of these problems is to instruct the Hardhat Network to always fork from a given block number. In your config, it looks like this:

Reviewer: ```suggestion
To avoid these problems always fork from a specific block number, like this:
```

```js
networks: {
  hardhat: {
    forking: {
      url: "https://eth-mainnet.alchemyapi.io/v2/<key>",
      blockNumber: 11095000
    }
  }
}
```

> This way, every time the Hardhat Network is used, it will fork the mainnet from that block, significantly improving its performance and making it more reproducible.

Reviewer: ```suggestion
```

The way to do this in Hardhat Network is by impersonating accounts. When you impersonate an account, you can send any transaction from it as if you had its private key. You can even send transactions from smart contract addresses.

To impersonate an account, you have to use the `hardhat_impersonateAccount` RPC method, with the address to impersonate as its parameter:

```tsx
await hre.network.provider.request({
  method: "hardhat_impersonateAccount",
  params: ["0x364d6D0333432C3Ac016Ca832fb8594A8cE43Ca6"]}
)
```

If at any point you want to stop doing it, you can use the `hardhat_stopImpersonatingAccount` method:

```tsx
await hre.network.provider.request({
  method: "hardhat_stopImpersonatingAccount",
  params: ["0x364d6D0333432C3Ac016Ca832fb8594A8cE43Ca6"]}
)
```

## Resetting the fork

You might want to reset the Hardhat Network without having to stop it, for example to start from the same state between different test fixtures. You can do this with the `hardhat_reset` method:

```ts
await network.provider.request({
  method: "hardhat_reset",
  params: [{
    forking: {

>       jsonRpcUrl: "https://eth-mainnet.alchemyapi.io/v2/<projectId>",

Reviewer 2: ```suggestion
      jsonRpcUrl: "https://eth-mainnet.alchemyapi.io/v2/<key>",
```
